In [1]:
import pandas as pd
import datetime

# 01. 유튜브 인기 동영상 데이터

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/youtube.csv", index_col=0)
df.head()

title             channelTitle  \
0                                          [신병] 물자창고                      장삐쭈   
1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV  RAIN's Official Channel   
2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.              서울시 · Seoul   
3                                       고기남자의 칠면조 파티             고기남자 MeatMan   
4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...                    스브스밥집   

   categoryId  view_count  likes  dislikes  comment_count  \
0          23     1893473  38249       730           8595   
1          10     2600864      0         0          20129   
2          29      347049   3564       120            178   
3          26      528458  15372       280           3470   
4          24      494904   3918       111           3142   

                  channelId trending_date2  
0  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 60399 entries, 0 to 60398
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   title           60399 non-null  object
 1   channelTitle    60399 non-null  object
 2   categoryId      60399 non-null  int64 
 3   view_count      60399 non-null  int64 
 4   likes           60399 non-null  int64 
 5   dislikes        60399 non-null  int64 
 6   comment_count   60399 non-null  int64 
 7   channelId       60399 non-null  object
 8   trending_date2  60399 non-null  object
dtypes: int64(5), object(4)
memory usage: 4.6+ MB


**Question 01. 인기 동영상 제작 횟수가 많은 채널 상위 10개명을 출력하라(날짜 기준, 중복 포함).**

In [4]:
df['trending_date2'] = pd.to_datetime(df['trending_date2'])
df['trending_date2'].head()

0   2021-01-01
1   2021-01-01
2   2021-01-01
3   2021-01-01
4   2021-01-01
Name: trending_date2, dtype: datetime64[ns]

In [5]:
# sol1
list(df.loc[df.channelId.isin(df.channelId.value_counts().head(10).index)].channelTitle.unique())

['장삐쭈',
 '총몇명',
 '파뿌리',
 '짤툰',
 '런닝맨 - 스브스 공식 채널',
 '엠뚜루마뚜루 : MBC 공식 종합 채널',
 'SPOTV',
 '채널 십오야',
 '이과장',
 'BANGTANTV']

In [6]:
# sol2
df[df['channelId'].isin(df['channelId'].value_counts().index[:10])]['channelTitle'].drop_duplicates().reset_index(drop=True)

0                      장삐쭈
1                      총몇명
2                      파뿌리
3                       짤툰
4          런닝맨 - 스브스 공식 채널
5    엠뚜루마뚜루 : MBC 공식 종합 채널
6                    SPOTV
7                   채널 십오야
8                      이과장
9                BANGTANTV
Name: channelTitle, dtype: object

**Question 03. 채널명을 바꾼 케이스가 있는지 확인하고 싶다. channelId의 경우
고윳값이므로 이를 통해 채널명을 한 번이라도 바꾼 채널의 개수를 구하여라.**

In [7]:
# sol1
target = df[['channelTitle', 'channelId']].drop_duplicates().groupby(['channelId']).size()
len(target[target>1])

71

In [8]:
# sol2
target = df[['channelTitle', 'channelId']].drop_duplicates()['channelId'].value_counts()
len(target[target>1])

71

**Question 05. 요일별 인기 영상들의 categoryId는 각각 몇 개씩인지 하나의 데이터프레임으로 표현하라.**

In [9]:
# sol1
df.groupby(df['trending_date2'].dt.day_name())['categoryId'].value_counts().unstack()

categoryId       1    2    10   15   17  19   20    22   23    24   25   26  \
trending_date2                                                                
Friday          243  120  833  187  633  90  283  1288  568  2976  444  369   
Monday          263  105  837  215  668  92  298  1373  594  3148  453  378   
Saturday        255  119  776  198  592  87  296  1289  570  3066  422  364   
Sunday          274   99  830  217  636  91  289  1337  556  3096  437  363   
Thursday        246  128  890  207  682  92  282  1341  560  2954  470  375   
Tuesday         257  119  894  208  708  89  285  1375  569  3084  452  394   
Wednesday       234  129  917  207  706  85  291  1333  566  3090  468  385   

categoryId       27   28  29  
trending_date2                
Friday          183  171  12  
Monday          205  160  10  
Saturday        183  173  10  
Sunday          199  167   9  
Thursday        194  166  13  
Tuesday         194  161  11  
Wednesday       212  165  12

In [10]:
# sol2
df.groupby([df['trending_date2'].dt.day_name(), 'categoryId'])['categoryId'].size().unstack()

categoryId       1    2    10   15   17  19   20    22   23    24   25   26  \
trending_date2                                                                
Friday          243  120  833  187  633  90  283  1288  568  2976  444  369   
Monday          263  105  837  215  668  92  298  1373  594  3148  453  378   
Saturday        255  119  776  198  592  87  296  1289  570  3066  422  364   
Sunday          274   99  830  217  636  91  289  1337  556  3096  437  363   
Thursday        246  128  890  207  682  92  282  1341  560  2954  470  375   
Tuesday         257  119  894  208  708  89  285  1375  569  3084  452  394   
Wednesday       234  129  917  207  706  85  291  1333  566  3090  468  385   

categoryId       27   28  29  
trending_date2                
Friday          183  171  12  
Monday          205  160  10  
Saturday        183  173  10  
Sunday          199  167   9  
Thursday        194  166  13  
Tuesday         194  161  11  
Wednesday       212  165  12

In [11]:
# sol3
target = df.groupby([df['trending_date2'].dt.day_name(), 'categoryId'], as_index=False).size()
target.pivot(index='trending_date2', columns='categoryId', values='size')

categoryId       1    2    10   15   17  19   20    22   23    24   25   26  \
trending_date2                                                                
Friday          243  120  833  187  633  90  283  1288  568  2976  444  369   
Monday          263  105  837  215  668  92  298  1373  594  3148  453  378   
Saturday        255  119  776  198  592  87  296  1289  570  3066  422  364   
Sunday          274   99  830  217  636  91  289  1337  556  3096  437  363   
Thursday        246  128  890  207  682  92  282  1341  560  2954  470  375   
Tuesday         257  119  894  208  708  89  285  1375  569  3084  452  394   
Wednesday       234  129  917  207  706  85  291  1333  566  3090  468  385   

categoryId       27   28  29  
trending_date2                
Friday          183  171  12  
Monday          205  160  10  
Saturday        183  173  10  
Sunday          199  167   9  
Thursday        194  166  13  
Tuesday         194  161  11  
Wednesday       212  165  12

**Question 06. 댓글 수로 (comment_count) 영상 반응에 대한 판단을 할 수 있다. view_count 대비 댓글 수가 가장 높은 영상을 확인하라(view_count 값이 0인 경우는 제외한다).**

In [12]:
target1 = df[df['view_count']!=0]
target2 = target1.copy()
target2['ratio'] = target1['comment_count'] / target1['view_count']
target2.sort_values(by='ratio').iloc[-1, :]['title']

'60분 동안 댓글이 달리지 않으면, 영상이 삭제됩니다. (챌린지)'

**Question 07. 댓글 수로 (comment_count) 영상 반응에 대한 판단을 할 수 있다. view_count 대비 댓글 수가 가장 낮은 영상을 확인하라(view_count, ratio 값이 0인 경우는 제외한다).**

In [13]:
# sol1
target1 = df[(df['view_count']!=0)]
target2 = target1.copy()
target2['ratio'] = target2['comment_count'] / target2['view_count']
target2[target2['ratio']!=0].sort_values(by='ratio').iloc[0, :]['title']

'Join the BTS #PermissiontoDance Challenge only on YouTube #Shorts'

In [14]:
# sol2
ratio = (df['comment_count'] / df['view_count']).dropna().sort_values()
df.iloc[ratio[ratio!=0].index[0]]['title']

'Join the BTS #PermissiontoDance Challenge only on YouTube #Shorts'

**Question 08. like 대비 dislike의 수가 가장 적은 영상은 무엇인가?(like, dislike 값이 0인경우는 제외한다).**

In [15]:
# sol1
ratio = (df['dislikes'] / df['likes']).dropna().sort_values()
df.iloc[ratio[ratio!=0].index[0]]['title']

'[줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰'

In [16]:
# sol2
target1 = df[(df['likes']!=0) & (df['dislikes']!=0)]
target2 = target1.copy()
target2['ratio'] = target2['dislikes'] / target2['likes']
target2.sort_values(by='ratio').iloc[0]['title']

'[줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰'

In [17]:
# sol3
target = df.loc[(df.likes!=0) & (df.dislikes!=0)]
num = (target['dislikes'] / target['likes']).sort_values().index[0]
df.iloc[num].title

'[줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰'

**Question 09. 가장 많은 트렌드 영상을 제작한 채널의 이름은 무엇인가?(날짜 기준, 중복 포함).**

In [18]:
df[df['channelId']==df['channelId'].value_counts().index[0]]['channelTitle'].unique()[0]

'짤툰'

**Question 10. 20회(20일)이상 인기 동영상 리스트에 포함된 동영상의 숫자는?**

In [19]:
(df[['title', 'channelId']].value_counts()>=20).sum()

40

# 02. 유튜브 공범 컨텐츠 동영상 데이터

In [20]:
channel = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/channelInfo.csv')
video = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/videoInfo.csv')

display(channel.head())
display(video.head())

channelid   subcnt    viewcnt  videocnt  \
0  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  410238653       736   
1  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       736   
2  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       735   
3  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       737   
4  UCkQCwnkQfgSuPTTnw_Y7v7w  1320000  412531322       737   

                    ct channelname  
0  2021-09-30 03:01:03         꽈뚜룹  
1  2021-09-30 09:01:03         꽈뚜룹  
2  2021-09-30 15:01:03         꽈뚜룹  
3  2021-09-30 21:01:03         꽈뚜룹  
4  2021-10-01 03:01:04         꽈뚜룹

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
0  c5JQp6xafqc  1667010    30474         706            0   6587   
1  c5JQp6xafqc  1669089    30495         707            0   6589   
2  c5JQp6xafqc  1674759    30522         711            0   6596   
3  c5JQp6xafqc  1677026    30555         712            0   6604   
4  c5JQp6xafqc  1681824    30585         713            0   6600   

                    ct videoname  
0  2021-10-10 15:20:03    공범 EP1  
1  2021-10-10 15:30:03    공범 EP1  
2  2021-10-10 15:40:02    공범 EP1  
3  2021-10-10 15:50:03    공범 EP1  
4  2021-10-10 16:00:03    공범 EP1

**Question 12. 수집된 각 video의 가장 최신화 된 날짜의 viewcount 값을 출력하라.**

In [21]:
# sol1
video['ct'] = pd.to_datetime(video['ct'])
target = video[['ct', 'videoname', 'viewcnt']].value_counts().reset_index().sort_values(by='ct').drop_duplicates('videoname', keep='last')
target[['ct', 'videoname', 'viewcnt']].sort_values(by='videoname').reset_index(drop=True)

ct videoname  viewcnt
0 2021-11-01 15:30:03    공범 EP1  3180532
1 2021-11-01 15:30:03    공범 EP2  2199328
2 2021-11-01 15:30:03    공범 EP3  1671294
3 2021-11-01 15:30:03    공범 EP4  1818493
4 2021-11-01 15:30:04    공범 EP5  1503435
5 2021-11-01 15:30:04    공범 EP6  1750222
6 2021-11-01 15:30:05    공범 EP7  1630200
7 2021-11-01 15:30:05    공범 EP8  1289088

In [22]:
# so12
video.sort_values(['videoname', 'ct']).drop_duplicates('videoname', keep='last')[['viewcnt', 'videoname', 'ct']].reset_index(drop=True)

viewcnt videoname                  ct
0  3180532    공범 EP1 2021-11-01 15:30:03
1  2199328    공범 EP2 2021-11-01 15:30:03
2  1671294    공범 EP3 2021-11-01 15:30:03
3  1818493    공범 EP4 2021-11-01 15:30:03
4  1503435    공범 EP5 2021-11-01 15:30:04
5  1750222    공범 EP6 2021-11-01 15:30:04
6  1630200    공범 EP7 2021-11-01 15:30:05
7  1289088    공범 EP8 2021-11-01 15:30:05

**Question 13. Channel 데이터 중 2021-10-03일 이후 각 채널의 처음 기록 됐던 구독자 수(subcnt)를 출력하라.**

In [23]:
channel['ct'] = pd.to_datetime(channel['ct'])

In [24]:
channel[(channel.ct>'2021-10-03')].sort_values(['ct', 'channelname']).drop_duplicates('channelname', keep='first')[['channelname', 'subcnt']]

channelname   subcnt
393            논리왕 전기   922000
139             야전삽짱재   257000
266        츄정ChuJeong   322000
901        김농밀의 농밀한 삶     7520
1028      릴펄 Lilpearl    10100
520   와글와글 WagleWagle    55000
647               조나단   215000
774           형사!탐정되다    14900
1276    Balming Tiger    54300
1155    곽토리 kwak tori   471000
12                꽈뚜룹  1330000

**Question 14. 각 채널의 2021-10-03 03:00:00 ~ 2021-11-01 15:00:00 까지 구독자 수(subcnt)의 증가량을 구하여라.**

In [25]:
# sol1
start = channel[channel['ct']>='2021-10-03 03:00:00']
end = channel[channel['ct']<='2021-11-01 15:00:00']

start_channel = start[['ct', 'channelname', 'subcnt']].drop_duplicates('channelname', keep='first')
end_channel = end[['ct',  'channelname', 'subcnt']].drop_duplicates('channelname', keep='last')

# start_channel.columns = ['ct', 'channelname', 'startcnt']
# end_channel.columns = ['ct', 'channelname', 'endcnt']

target = pd.merge(start_channel, end_channel, on='channelname')
target['diff'] = target['subcnt_y']-target['subcnt_x']
target[['channelname', 'diff']].sort_values(by='diff').sort_index()

channelname   diff
0               꽈뚜룹  70000
1             야전삽짱재  11000
2        츄정ChuJeong   1000
3            논리왕 전기 -11000
4   와글와글 WagleWagle      0
5               조나단  12000
6           형사!탐정되다  10300
7        김농밀의 농밀한 삶   1540
8       릴펄 Lilpearl  11000
9     곽토리 kwak tori  -2000
10    Balming Tiger   2500

In [26]:
# sol2
end = channel.loc[channel.ct.dt.strftime('%Y-%m-%d %H')=='2021-11-01 15']
start = channel.loc[channel.ct.dt.strftime('%Y-%m-%d %H')=='2021-10-03 03']

end_df = end[['channelname', 'subcnt']].reset_index(drop=True)
start_df = start[['channelname', 'subcnt']].reset_index(drop=True)

end_df.columns = ['channelname', 'end_sub']
start_df.columns = ['channelname', 'start_sub']

tt = pd.merge(start_df, end_df)
tt['del'] = tt['end_sub'] - tt['start_sub']
tt[['channelname', 'del']]

channelname    del
0               꽈뚜룹  70000
1             야전삽짱재  11000
2        츄정ChuJeong   1000
3            논리왕 전기 -11000
4   와글와글 WagleWagle      0
5               조나단  12000
6           형사!탐정되다  10300
7        김농밀의 농밀한 삶   1540
8       릴펄 Lilpearl  11000
9     곽토리 kwak tori  -2000
10    Balming Tiger   2500

**Question 15. 공범 EP1의 비디오 정보 데이터 중 수집 간격이 5분 이하, 20분 이상인 데이터 구간(해당 시점 전, 후)의 시각을 모두 출력하라.**

In [27]:
video.head()

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
0  c5JQp6xafqc  1667010    30474         706            0   6587   
1  c5JQp6xafqc  1669089    30495         707            0   6589   
2  c5JQp6xafqc  1674759    30522         711            0   6596   
3  c5JQp6xafqc  1677026    30555         712            0   6604   
4  c5JQp6xafqc  1681824    30585         713            0   6600   

                   ct videoname  
0 2021-10-10 15:20:03    공범 EP1  
1 2021-10-10 15:30:03    공범 EP1  
2 2021-10-10 15:40:02    공범 EP1  
3 2021-10-10 15:50:03    공범 EP1  
4 2021-10-10 16:00:03    공범 EP1

In [28]:
target = video[video['videoname'].str.contains('EP1')].sort_values(by='ct')
target[(target.ct.diff(1) >= datetime.timedelta(minutes=20)) |\
       (target.ct.diff(1) <= datetime.timedelta(minutes=5))].index
# Int64Index([721, 722, 1636], dtype='int64')
target[target.index.isin([720, 721, 722, 723, 1635, 1636, 1637])]

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
720   c5JQp6xafqc  2153676    34104         830            0   7120   
721   c5JQp6xafqc  2228250    34559         849            0   7191   
722   c5JQp6xafqc  2228250    34559         849            0   7191   
723   c5JQp6xafqc  2228669    34566         849            0   7194   
1635  c5JQp6xafqc  2706691    37893         987            0   7410   
1636  c5JQp6xafqc  2707933    37901         988            0   7410   
1637  c5JQp6xafqc  2708448    37903         988            0   7410   

                      ct videoname  
720  2021-10-12 19:20:03    공범 EP1  
721  2021-10-13 09:41:37    공범 EP1  
722  2021-10-13 09:41:37    공범 EP1  
723  2021-10-13 09:50:03    공범 EP1  
1635 2021-10-19 17:50:02    공범 EP1  
1636 2021-10-19 18:20:03    공범 EP1  
1637 2021-10-19 18:30:03    공범 EP1

**Question 16. 각 에피소드의 시작 날짜(년-월-일)를 에피소드 이름과 묶어 데이터프레임으로 만들고 출력하라.**

In [29]:
# sol1
target = video.copy()
target['ct'] = target['ct'].dt.date
target.sort_values(['videoname', 'ct']).drop_duplicates('videoname', keep='first')[['videoname', 'ct']]

videoname          ct
10       공범 EP1  2021-10-07
3496     공범 EP2  2021-10-09
6696     공범 EP3  2021-10-14
9264     공범 EP4  2021-10-16
11544    공범 EP5  2021-10-21
13661    공범 EP6  2021-10-23
13106    공범 EP7  2021-10-28
14935    공범 EP8  2021-10-30

In [30]:
# sol2
start_date = video.sort_values(['ct', 'videoname']).drop_duplicates('videoname')[['ct', 'videoname']]
start_date['date'] = start_date.ct.dt.date
start_date[['date', 'videoname']]

date videoname
10     2021-10-07    공범 EP1
3496   2021-10-09    공범 EP2
6696   2021-10-14    공범 EP3
9264   2021-10-16    공범 EP4
11544  2021-10-21    공범 EP5
13661  2021-10-23    공범 EP6
13106  2021-10-28    공범 EP7
14935  2021-10-30    공범 EP8

**Question 17. 공범 컨텐츠의 경우 19:00시에 공개 되는것으로 알려져있다. 공개된 날의 21시의 viewcnt, ct, videoname 으로 구성된 데이터프레임을 viewcnt를 내림차순으로 정렬하여 출력하라.**

In [31]:
# sol1
target = video[video['ct'].dt.hour==21]
target.sort_values(['ct', 'videoname']).drop_duplicates('videoname', keep='first')[['viewcnt', 'ct', 'videoname']].sort_values(by='viewcnt', ascending=False).reset_index(drop=True)

viewcnt                  ct videoname
0   264029 2021-10-30 21:00:08    공범 EP8
1   252032 2021-10-28 21:00:04    공범 EP7
2   217674 2021-10-16 21:00:04    공범 EP4
3   213899 2021-10-23 21:00:06    공범 EP6
4   201179 2021-10-21 21:00:04    공범 EP5
5   148144 2021-10-09 21:00:03    공범 EP2
6   147183 2021-10-14 21:00:04    공범 EP3
7   117340 2021-10-07 21:00:03    공범 EP1

In [32]:
# sol2
video['time'] = video.ct.dt.hour
video.loc[video['time']==21].sort_values(['videoname', 'ct']).drop_duplicates('videoname')[['viewcnt', 'ct', 'videoname']].sort_values(by='viewcnt', ascending=False).reset_index(drop=True)

viewcnt                  ct videoname
0   264029 2021-10-30 21:00:08    공범 EP8
1   252032 2021-10-28 21:00:04    공범 EP7
2   217674 2021-10-16 21:00:04    공범 EP4
3   213899 2021-10-23 21:00:06    공범 EP6
4   201179 2021-10-21 21:00:04    공범 EP5
5   148144 2021-10-09 21:00:03    공범 EP2
6   147183 2021-10-14 21:00:04    공범 EP3
7   117340 2021-10-07 21:00:03    공범 EP1

**Question 18. video 정보의 가장 최근 데이터들에서 각 에피소드의 싫어요/좋아요 비율을 ratio 컬럼으로 만들고 videoname, ratio로 구성된 데이터프레임을 ratio를 오름차순으로 정렬하라.**

In [33]:
target = video[['videoname', 'likecnt', 'dislikecnt', 'ct']].sort_values(by='ct').drop_duplicates('videoname', keep='last')
target['ratio'] = target['dislikecnt'] / target['likecnt']
target[['videoname', 'ratio']].sort_values(by='ratio').reset_index(drop=True)

videoname     ratio
0    공범 EP7  0.013636
1    공범 EP6  0.015881
2    공범 EP4  0.017971
3    공범 EP5  0.021128
4    공범 EP3  0.022360
5    공범 EP1  0.027118
6    공범 EP2  0.029844
7    공범 EP8  0.141357

**Question 19. 2021-11-01 00:00:00 ~ 15:00:00까지 에피소드별 viewcnt의 증가량을 데이터프레임으로 만드시오.**

In [34]:
# sol1
target = video[(video['ct']>='2021-11-01 00:00:00') & (video['ct']<='2021-11-01 15:00:00')]
start = target[['videoname', 'ct', 'viewcnt']].sort_values(by='ct').drop_duplicates('videoname', keep='first')
end = target[['videoname', 'ct', 'viewcnt']].sort_values(by='ct').drop_duplicates('videoname', keep='last')
start.columns = ['videoname', 'ct', 'start_cnt']
end.columns = ['videoname', 'ct', 'end_cnt']
result = pd.merge(start, end, on='videoname')
result['viewcnt'] = result['end_cnt'] - result['start_cnt']
result[['videoname', 'viewcnt']].sort_values(by='videoname')

videoname  viewcnt
0    공범 EP1    13298
1    공범 EP2    10300
2    공범 EP3     9927
4    공범 EP4     9824
5    공범 EP5    10824
3    공범 EP6    14141
7    공범 EP7    26949
6    공범 EP8    89147

In [35]:
# sol2
start = '2021-11-01 00:00:00'
end = '2021-11-01 15:00:00'

target = video.loc[(video['ct']>=start) & (video['ct']<=end)].reset_index(drop=True)

def check(x):
    result = max(x) - min(x)
    return result

target[['videoname', 'viewcnt']].groupby('videoname').agg(check)

viewcnt
videoname         
 공범 EP1      13298
 공범 EP2      10300
 공범 EP3       9927
 공범 EP4       9824
 공범 EP5      10824
 공범 EP6      14141
 공범 EP7      26949
 공범 EP8      89147

# 03. 월드컵 출전 선수 골 기록 데이터

In [36]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/worldcup/worldcupgoals.csv')
df.head()

Player  Goals                Years  Country
0  Miroslav Klose     16  2002-2006-2010-2014  Germany
1         Ronaldo     15       1998-2002-2006   Brazil
2     Gerd Muller     14            1970-1974  Germany
3   Just Fontaine     13                 1958   France
4            Pele     12  1958-1962-1966-1970   Brazil

**Question 21. 주어진 전체 기간의 나라별 골 득점수 상위 5개 국가와 그 득점수를 데이터프레임 형태로 출력하라.**

In [37]:
df[['Country', 'Goals']].groupby('Country').sum().sort_values(by='Goals', ascending=False).head()

Goals
Country         
Brazil       228
Germany      226
Argentina    135
Italy        123
France       115

**Question 22. 주어진 전체 기간 동안 골 득점을 한 선수가 가장 많은 나라 상위 5개 국가와 그 선수 숫자를 데이터프레임 형식으로 출력하라.**

In [38]:
df.groupby(['Country'])['Player'].size().to_frame('size').sort_values(by='size', ascending=False).head(5)

size
Country        
Brazil       81
Germany      78
Italy        60
Argentina    59
France       58

**Quesiton 23. Years 컬럼은 년도-년도 형식으로 구성되어있고, 각 년도는 4자리 숫자이다. 년도 표기가 4자리 숫자로 안된 케이스가 존재한다. 해당 건은 몇 건인지 출력하라.**

In [39]:
df['yearLst'] = df['Years'].str.split('-')

def checkFour(x):
    for value in x:
        if len(str(value))!=4:
            return False
    else:
        return True

df['check'] = df['yearLst'].apply(checkFour)
len(df[df['check']==False])

45

In [40]:
df['yearLst'].str.len()

0       4
1       3
2       2
3       1
4       4
       ..
1290    1
1291    2
1292    1
1293    1
1294    2
Name: yearLst, Length: 1295, dtype: int64

# 04. 서울시 따릉이 이용 정보 데이터

In [41]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/bicycle/seoul_bi.csv')
df.head()

대여일자  대여시간  대여소번호 대여구분코드 성별 연령대코드  이용건수    운동량   탄소량     이동거리  사용시간
0  2021-06-01     0   3541    정기권  F  ~10대     1   0.00  0.00     0.00     8
1  2021-06-01     0    765    정기권  F  ~10대     1  27.21  0.35  1526.81    19
2  2021-06-01     0   2637    정기권  F  ~10대     1  41.40  0.37  1608.56    18
3  2021-06-01     0   2919    정기권  F  ~10대     1   0.00  0.00     0.00    75
4  2021-06-01     0    549    정기권  F  ~10대     1  13.04  0.17   731.55     6

**Question 32. 각 일자의 요일을 표기하고 (‘Monday’ ~’Sunday’) ‘day_name’ 컬럼을 추가하고 이를 이용하여 요일별 이용 횟수의 총합을 데이터프레임으로 출력하라.**

In [42]:
# 이용건수에 따라서 계산
df['대여일자'] = pd.to_datetime(df['대여일자'])
df['day_name'] = df['대여일자'].dt.day_name()
df.groupby('day_name')['이용건수'].sum().to_frame('cnt')

cnt
day_name         
Friday     120716
Monday     119271
Saturday   110900
Sunday      99297
Thursday    60338
Tuesday    114492
Wednesday  122653

In [43]:
# 요일별 건수에 따라 계산
df['day_name'].value_counts().to_frame()

day_name
Wednesday    110607
Friday       108877
Monday       107568
Tuesday      103704
Saturday     101299
Sunday        90918
Thursday      55977

**Quesiton 33. 요일별 가장 많이 이용한 대여소의 이용횟수와 대여소 번호를 데이터프레임으로 출력하라.**

In [44]:
# sol1
target = df.groupby(['day_name', '대여소번호']).size().reset_index().sort_values(by=0, ascending=False).drop_duplicates('day_name', keep='first')
target.reset_index(drop=True).rename(columns={0: 'cnt'})

day_name  대여소번호  cnt
0   Saturday    502  378
1     Sunday    502  372
2  Wednesday    502  282
3     Friday    502  277
4    Tuesday    502  267
5     Monday    502  242
6   Thursday   2715  137

In [45]:
# sol2
target = df.groupby(['day_name', '대여소번호']).size().to_frame('size').sort_values(['day_name', 'size'], ascending=False).reset_index()
target.drop_duplicates('day_name', keep='first').reset_index(drop=True)

day_name  대여소번호  size
0  Wednesday    502   282
1    Tuesday    502   267
2   Thursday   2715   137
3     Sunday    502   372
4   Saturday    502   378
5     Monday    502   242
6     Friday    502   277

**Question 34. 나이대별 대여구분 코드의 (일일권/전체횟수) 비율을 구한 후 가장 높은 비율을 가지는 나이대를 확인하라. 일일권의 경우 일일권과 일일권(비회원)을 모두 포함하라.**

In [46]:
daily = df[df.대여구분코드.isin(['일일권', '일일권(비회원)'])].연령대코드.value_counts().sort_index()
total = df.연령대코드.value_counts().sort_index()
ratio = (daily/total).sort_values(ascending=False)
print(ratio)
print(ratio.index[0])

~10대    0.312525
20대     0.239913
30대     0.166702
40대     0.110585
50대     0.065414
70대~    0.048962
60대     0.042462
Name: 연령대코드, dtype: float64
~10대


**Question 35. 연령대별 평균 이동거리를 구하여라.**

In [47]:
df[['연령대코드', '이동거리']].groupby(['연령대코드']).mean()

이동거리
연령대코드             
20대    3211.890552
30대    3341.443859
40대    3514.857416
50대    3593.668100
60대    3538.145737
70대~   3085.039641
~10대   2634.426279

**Question 36. 연령대 코드가 20대인 데이터를 추출하고, 이동거리값이 추출한 데이터의 이동거리값의 평균 이상인 데이터를 추출한다. 최종 추출된 데이터를 대여일자, 대여소 번호 순서로 내림차순 정렬 후 1행부터 200행까지의 탄소량의 평균을 소숫점 3째 자리까지 구하여라.**

In [48]:
# sol1
target1 = df[(df['연령대코드']=='20대') & (df['이동거리']>=df['이동거리'].mean())]
target2 = target1.copy()
target2['탄소량'] = target2['탄소량'].astype('float')
result = target2.sort_values(by=['대여일자', '대여소번호'], ascending=False).iloc[:200]['탄소량']
round(result.sum()/len(result), 3)

1.643

In [49]:
df['탄소량'] = pd.to_numeric(df['탄소량'], errors='coerce')
# df['탄소량']에 바로 astype('float')하면 먹히지 않고 copy()해서 쓰면 된다
target1 = df[(df['연령대코드']=='20대') & (df['이동거리']>=df['이동거리'].mean())]
target2 = target1.sort_values(by=['대여일자', '대여소번호'], ascending=False).iloc[:200]['탄소량']
round(target2.sum()/len(target2), 3)

1.643

**Question 38. 평일(월~금) 출근 시간대(오전 6, 7, 8시)의 대여소별 이용 횟수를 구해서 데이터프레임 형태로 표현한 후, 대여시간별 이용 횟수의 상위 3개 대여소와 이용 횟수를 출력하라.**

In [50]:
target = df[(df['대여일자'].dt.weekday.isin([0, 1, 2, 3, 4])) & (df['대여시간'].isin([6, 7, 8]))]
result = target.groupby(['대여시간', '대여소번호']).size().to_frame('이용 횟수')
result.sort_values(['대여시간', '이용 횟수'], ascending=False).groupby('대여시간').head(3)

이용 횟수
대여시간 대여소번호       
8    2701     119
     646      115
     1152      92
7    259      104
     230       77
     726       77
6    2744      45
     1125      40
     1028      36

# 05. 전세계 행복도 지표 데이터

In [51]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/happy2/happiness.csv',encoding='utf-8')
df.head()

행복랭킹          나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식  \
0     1      Finland  7.769  1.340  1.587  0.986    0.596  0.153     0.393   
1     2      Denmark  7.600  1.383  1.573  0.996    0.592  0.252     0.410   
2     3       Norway  7.554  1.488  1.582  1.028    0.603  0.271     0.341   
3     4      Iceland  7.494  1.380  1.624  1.026    0.591  0.354     0.118   
4     5  Netherlands  7.488  1.396  1.522  0.999    0.557  0.322     0.298   

     년도  
0  2019  
1  2019  
2  2019  
3  2019  
4  2019

**Question 45. 2019년도 데이터들만 추출하여 각 변수 간 상관계수를 구하고 내림차순으로 정렬한 후, 상위 5개를 데이터 프레임으로 출력하라. 컬럼명은 v1, v2, corr으로 표시하라.**

In [52]:
# sol1
target = df[df['년도']==2019].corr(numeric_only=True).unstack().to_frame().reset_index().rename(columns={0: 'corr',
                                                                                                          'level_0': 'v1',
                                                                                                          'level_1': 'v2'})
target[(target['v1']!=target['v2'])].drop_duplicates('corr').sort_values(by='corr', ascending=False).head(5)

v1     v2      corr
22  상대GDP  행복기대치  0.835462
11     점수  상대GDP  0.793883
13     점수  행복기대치  0.779883
12     점수  사회적지원  0.777058
21  상대GDP  사회적지원  0.754906

In [53]:
# sol2
target = df[df['년도']==2019].corr(numeric_only=True).unstack().to_frame().reset_index().dropna()
result = target[target[0]!=1].sort_values(0, ascending=False).drop_duplicates(0)
answer = result.head(5).reset_index(drop=True)
answer.columns = ['v1', 'v2', 'corr']
answer

v1     v2      corr
0  행복기대치  상대GDP  0.835462
1  상대GDP     점수  0.793883
2  행복기대치     점수  0.779883
3  사회적지원     점수  0.777058
4  사회적지원  상대GDP  0.754906

**Question 46. 년도별 하위 행복점수의 하위 5개 국가의 평균 행복점수를 구하여라.**

In [54]:
# sol1
df.sort_values(by='점수').groupby('년도').head(5).groupby('년도')['점수'].mean(numeric_only=True)

년도
2018    3.1800
2019    3.1408
Name: 점수, dtype: float64

In [55]:
# sol2
df.groupby('년도').tail(5).groupby('년도')['점수'].mean(numeric_only=True)

년도
2018    3.1800
2019    3.1408
Name: 점수, dtype: float64

**Question 50. 2018년, 2019년 모두 기록이 있는 나라들 중 연도별 행복점수가 가장 증가한 나라와 그 증가 수치는?**

In [56]:
cnt = df['나라명'].value_counts()
target = cnt[cnt>=2].index

df2 = df.copy()
multiple = df2[df2['나라명'].isin(target)].reset_index(drop=True)
multiple.loc[multiple['년도']==2018, '점수'] = multiple[multiple['년도']==2018]['점수'].values * (-1)
# 위의 코드에서 2018년도 값을 -로 대입해 준 것
# 그래야 groupby 이후 sum 했을 때 차이값이 계산되므로
result = multiple.groupby('나라명').sum()['점수'].sort_values().to_frame().iloc[-1]
result

점수    0.87
Name: Burundi, dtype: float64

In [57]:
df[df['나라명']=='Malaysia']

행복랭킹       나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식  \
79     80  Malaysia  5.339  1.221  1.171  0.828    0.508  0.260     0.024   
190    35  Malaysia  6.322  1.161  1.258  0.669    0.356  0.311     0.059   

       년도  
79   2019  
190  2018

In [58]:
df[df['나라명']=='Burundi']

행복랭킹      나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식    년도
144   145  Burundi  3.775  0.046  0.447  0.380    0.220  0.176     0.180  2019
311   156  Burundi  2.905  0.091  0.627  0.145    0.065  0.149     0.076  2018

In [59]:
multiple.loc[multiple['년도']==2019, '점수']

0      7.769
1      7.600
2      7.554
3      7.494
4      7.488
       ...  
147    3.334
148    3.231
149    3.203
150    3.083
151    2.853
Name: 점수, Length: 152, dtype: float64

# 06. 지역구 에너지 소비량 데이터

In [60]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/consum/Tetuan%20City%20power%20consumption.csv')
df.head()

DateTime  Temperature  Humidity  Wind Speed  general diffuse flows  \
0  1/1/2017 0:00        6.559      73.8       0.083                  0.051   
1  1/1/2017 0:10        6.414      74.5       0.083                  0.070   
2  1/1/2017 0:20        6.313      74.5       0.080                  0.062   
3  1/1/2017 0:30        6.121      75.0       0.083                  0.091   
4  1/1/2017 0:40        5.921      75.7       0.081                  0.048   

   diffuse flows  Zone 1 Power Consumption  Zone 2  Power Consumption  \
0          0.119               34055.69620                16128.87538   
1          0.085               29814.68354                19375.07599   
2          0.100               29128.10127                19006.68693   
3          0.096               28228.86076                18361.09422   
4          0.085               27335.69620                17872.34043   

   Zone 3  Power Consumption  
0                20240.96386  
1                20131.08434  
2                19668.43373  
3                18899.27711  
4                18442.40964

**Question 52. 3월달의 시간대별 온도의 평균들 중 가장 낮은 시간대의 온도를 출력하라.**

In [61]:
df['DateTime'] = pd.to_datetime(df['DateTime'])
target = df[df['DateTime'].dt.month==3]
# result = target.groupby(target['DateTime'].dt.hour)['Temperature'].mean().to_frame().sort_values(by='Temperature').values[0]
result = target.groupby(target['DateTime'].dt.hour)['Temperature'].mean().sort_values().values[0]
print(result)

11.506612903225806


**Question 56. Temperature의 값이 10 미만의 경우 A, 10 이상 20 미만의 경우 B, 20 이상 30 미만의 경우 C, 그 외의 경우 D라고 할 때 각 단계의 데이터 숫자를 구하여라.**

In [62]:
def checkTem(x):
    if x<10:
        return 'A'
    elif x<20:
        return 'B'
    elif x<30:
        return 'C'
    else:
        return 'D'
    
df['Temperature'].apply(checkTem).value_counts()

B    26993
C    21105
A     2874
D     1444
Name: Temperature, dtype: int64

# 07. 포켓몬 정보 데이터

In [63]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/pok/Pokemon.csv')
df.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

**Question 66. 각 Generation의 Attack으로 오름차순 정렬 시 상위 3개 데이터(18개) Attack의 전체 평균을 구하여라**

In [64]:
df.sort_values(by=['Generation', 'Attack']).groupby('Generation').head(3)['Attack'].mean()

19.5

**Question 70. 한 번씩만 존재하는 (Type1, Type2)의 쌍을 각 세대(Generation)는 각각 몇 개씩 가지고 있는가?**

In [65]:
target1 = df[['Type 1', 'Type 2']].value_counts()
target2 = target1[target1==1]

lst = []
for value in target2.reset_index().values:
    t1 = value[0]
    t2 = value[1]
    
    sp = df[(df['Type 1']==t1) & (df['Type 2']==t2)]
    lst.append(sp)
    
result = pd.concat(lst).reset_index(drop=True)['Generation'].value_counts().sort_index()
result

1     1
2     4
3     5
4    13
5     7
6     9
Name: Generation, dtype: int64

# 08. 기온 강수량 데이터

In [66]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/weather/weather2.csv")
df.head()

time  이화동강수  이화동기온  수영동강수  수영동기온
0  2020-01-01 00:00:00    0.1   -3.9    0.0    2.5
1  2020-01-01 01:00:00    0.0   -3.1    0.0    3.4
2  2020-01-01 02:00:00    0.0   -1.8    0.0    3.8
3  2020-01-01 03:00:00    0.0   -0.8    0.0    4.6
4  2020-01-01 04:00:00    0.0   -0.1    0.0    5.1

**Question 81. 여름철(6월, 7월, 8월) 이화동이 수영동보다 높은 기온을 가진 시간대는 몇 개인가?**

In [67]:
df['time'] = pd.to_datetime(df['time'])
target = df[df['time'].dt.month.isin([6, 7, 8])]
len(target[target['이화동기온']>target['수영동기온']]['time'].dt.hour)

1415

# 09. 핸드폰 가격 예측 데이터

In [68]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/mobile/train.csv")
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

**Question 90. price_range의 각 value를 그룹핑하여 각 그룹의 n_cores 의 빈도가 가장 높은 value와 그 빈도수를 구하여라.**

In [69]:
df[['price_range', 'n_cores']].value_counts().reset_index().sort_values(by=['price_range', 0], ascending=False).drop_duplicates('price_range', keep='first')

price_range  n_cores   0
3            3        5  70
2            2        4  73
0            1        4  76
6            0        2  69

In [70]:
df[['price_range', 'n_cores']].groupby(['price_range', 'n_cores']).size().sort_values(0).groupby(level=0).tail(1)

C:\Users\futur\AppData\Local\Temp\ipykernel_18752\2738451872.py:1: FutureWarning: In a future version of pandas all arguments of Series.sort_values will be keyword-only.
  df[['price_range', 'n_cores']].groupby(['price_range', 'n_cores']).size().sort_values(0).groupby(level=0).tail(1)


price_range  n_cores
0            2          69
3            5          70
2            4          73
1            1          76
dtype: int64

# 10. 비행 탑승 경험 만족도 데이터

In [71]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/airline/train.csv")
df.head()

id  Gender   Customer Type  Age   Type of Travel     Class  \
0   55882    Male  Loyal Customer   42  Business travel  Business   
1   31906  Female  Loyal Customer   31  Business travel  Business   
2   85481    Male  Loyal Customer   51  Personal Travel       Eco   
3  102134    Male  Loyal Customer   44  Business travel  Business   
4   31580    Male  Loyal Customer   33  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0              473                      1                                  1   
1             3271                      4                                  4   
2              214                      5                                  0   
3             3097                      1                                  1   
4             3286                      1                                  1   

   Ease of Online booking  ...  Inflight entertainment  On-board service  \
0                       3  ...                       5                 5   
1                       4  ...                       3                 3   
2                       5  ...                       3                 4   
3                       3  ...                       4                 4   
4                       1  ...                       4                 2   

   Leg room service  Baggage handling  Checkin service  Inflight service  \
0                 5                 5                3                 5   
1                 5                 5                3                 4   
2                 4                 4                4                 4   
3                 4                 4                3                 4   
4                 5                 2                5                 1   

   Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  \
0            3                           3                       0.0   
1            3                           1                       5.0   
2            3                           2                       0.0   
3            5                           0                       0.0   
4            4                           0                       0.0   

   satisfaction  
0     satisfied  
1     satisfied  
2     satisfied  
3     satisfied  
4     satisfied  

[5 rows x 24 columns]

**Question 92. Arrival Delay in Minutes 컬럼이 결측치인 데이터 중, ‘neutral or dissatisfied’ 보다 ‘satisfied’의 수가 더 높은 Class는 어디인가?**

In [72]:
target = df.loc[df['Arrival Delay in Minutes'].isnull()].groupby(['Class', 'satisfaction'], as_index=False).size().pivot(index='Class', columns='satisfaction')
target[target['size']['neutral or dissatisfied'] < target['size']['satisfied']]

size          
satisfaction neutral or dissatisfied satisfied
Class                                         
Business                          36        76

# 11. 의료 비용 예측 데이터

In [73]:
train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/MedicalCost/train.csv")
train.head()

age     sex     bmi  children smoker     region      charges
0   64  female  39.330         0     no  northeast  14901.51670
1   47  female  27.830         0    yes  southeast  23065.42070
2   52  female  33.300         2     no  southwest  10806.83900
3   33  female  22.135         1     no  northeast   5354.07465
4   30    male  31.400         1     no  southwest   3659.34600

**Question 94. 흡연자와 비흡연자 각각 charges의 상위 10% 그룹의 평균의 차이는?**

In [74]:
# sol1
target1 = train[train['smoker']=='yes']
target2 = train[train['smoker']=='no']
target1[target1['charges']>=target1['charges'].quantile(0.9)]['charges'].mean() - target2[target2['charges']>=target2['charges'].quantile(0.9)]['charges'].mean()

29297.954548156144

In [75]:
# sol2
sy = train[(train['smoker']=='yes') & (train['charges']>=train[train['smoker']=='yes']['charges'].quantile(0.9))]
sn = train[(train['smoker']=='no') & (train['charges']>=train[train['smoker']=='no']['charges'].quantile(0.9))]
sy['charges'].mean() - sn['charges'].mean()

29297.954548156144

# 12. 킹카운티 주거지 가격 예측 문제 데이터

In [76]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/kingcountyprice//train.csv")
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  8961990160  20150413T000000  567500.0         3        2.5         2080   
1  9455200205  20140604T000000  525000.0         3        2.0         1540   
2  7853220670  20140918T000000  540000.0         3        2.5         2860   
3  3298201170  20141110T000000  350000.0         3        1.0          940   
4  7972604355  20140521T000000  218000.0         3        1.0         1020   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      4556     2.0           0     0  ...      8        2080              0   
1      7800     1.0           0     0  ...      8        1540              0   
2      8935     2.0           0     0  ...      8        2860              0   
3      7811     1.0           0     0  ...      6         940              0   
4      7874     1.0           0     0  ...      7        1020              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1999             0    98074  47.6036 -122.014           1530   
1      2004             0    98125  47.7041 -122.288           1510   
2      2004             0    98065  47.5336 -121.855           2650   
3      1959             0    98008  47.6195 -122.118           1180   
4      1956             0    98106  47.5175 -122.346           1290   

   sqft_lot15  
0        5606  
1        7800  
2        6167  
3        7490  
4        7320  

[5 rows x 21 columns]

**Question 95. bedrooms의 빈도가 가장 높은 값을 가지는 데이터들의 price의 상위 10%와 하위 10%값의 차이를 구하여라.**

In [77]:
t1 = df[df['bedrooms']==df['bedrooms'].value_counts().sort_values().index[-1]]['price'].quantile(0.9)
t9 = df[df['bedrooms']==df['bedrooms'].value_counts().sort_values().index[-1]]['price'].quantile(0.1)
t1 - t9

505500.0

# 13. 대학원 입학 가능성 데이터

In [78]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/admission/train.csv")
df.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0         246        328          110                  4  4.0  2.5  9.02   
1         100        323          113                  3  4.0  4.0  8.88   
2          79        296           95                  2  3.0  2.0  7.54   
3          53        334          116                  4  4.0  3.0  8.00   
4         444        321          114                  5  4.5  4.5  9.16   

   Research  Chance of Admit  
0         1             0.81  
1         1             0.79  
2         1             0.44  
3         1             0.78  
4         1             0.87

**Question 96. Serial No. 컬럼을 제외하고 ‘Chance of Admit’을 종속변수, 나머지 변수를 독립변수라 할때, 랜덤포레스트를 통해 회귀 예측할 때 변수 중요도 값을 출력하라(시드값에 따라 순서는 달라질수 있음).**

In [79]:
df.drop('Serial No.', axis=1, inplace=True)

In [80]:
from sklearn.ensemble import RandomForestRegressor

y = df.pop('Chance of Admit')
x = df
rf = RandomForestRegressor()
model = rf.fit(x, y)
pd.DataFrame({'importance': model.feature_importances_, 'columns': x.columns})

importance            columns
0    0.080255          GRE Score
1    0.036690        TOEFL Score
2    0.012836  University Rating
3    0.024850                SOP
4    0.025159                LOR
5    0.813016               CGPA
6    0.007195           Research

In [81]:
from sklearn.ensemble import RandomForestRegressor
# df.drop('Serial No.', axis=1, inplace=True)
# y = df.pop('Chance of Admit')
# x = df
rf = RandomForestRegressor()
model = rf.fit(x, y)
pd.DataFrame({'importance': model.feature_importances_}, x.columns).sort_values(by='importance', ascending=False)

importance
CGPA                 0.805837
GRE Score            0.092100
TOEFL Score          0.031909
SOP                  0.024882
LOR                  0.023766
University Rating    0.013506
Research             0.008000

# 14. 레드 와인 퀄리티 예측 데이터

In [82]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/redwine/train.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            5.0              0.38         0.01             1.6      0.048   
1            5.0              0.42         0.24             2.0      0.060   
2            7.1              0.36         0.30             1.6      0.080   
3            7.6              0.29         0.49             2.7      0.092   
4            7.7              0.51         0.28             2.1      0.087   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 26.0                  60.0  0.99084  3.70       0.75   
1                 19.0                  50.0  0.99170  3.72       0.74   
2                 35.0                  70.0  0.99693  3.44       0.50   
3                 25.0                  60.0  0.99710  3.31       0.61   
4                 23.0                  54.0  0.99800  3.42       0.74   

   alcohol  quality  
0     14.0        6  
1     14.0        8  
2      9.4        5  
3     10.1        6  
4      9.2        5

**Question 97. quality 값이 3인 그룹과 8인 그룹의 컬럼별 독립변수의 표준편차 값의 차이를 구할때 그값이 가장 큰 컬럼명을 구하여라.**

In [83]:
# sol1
q3 = df[df['quality']==3].std(axis=0).reset_index().rename(columns={0: 'q3std'})
q8 = df[df['quality']==8].std(axis=0).reset_index().rename(columns={0: 'q8std'})
target = pd.merge(q3, q8)
target['diff'] = abs(target['q3std'] - target['q8std'])
target[target['diff']==max(target['diff'])].iloc[0, 0]

'total sulfur dioxide'

In [84]:
# sol2
abs((df.loc[df.quality==8].std() - df.loc[df.quality==3].std())).sort_values().index[-1]

'total sulfur dioxide'

# 15. 현대 차량 가격 분류 문제 데이터

In [85]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv")
df.head()

model  year  price transmission  mileage fuelType  tax(£)   mpg  \
0   Tucson  2016  12795       Manual    36263   Diesel      30  61.7   
1      I10  2012   3995       Manual    22089   Petrol      20  61.4   
2      I30  2019  24000       Manual      555   Petrol     145  34.0   
3   Tucson  2017  12995       Manual    32147   Diesel      30  61.7   
4   Tucson  2018  14991    Semi-Auto    32217   Diesel     150  57.6   

   engineSize  
0         1.7  
1         1.2  
2         2.0  
3         1.7  
4         1.7

**Question 101. 정보(row수)가 가장 많은 상위 3차종의 price값의 각 평균값은?**

In [86]:
df[df['model'].isin(df['model'].value_counts().index[:3])].groupby(['model'])['price'].mean()

model
 I10        7646.137891
 I30       11651.821759
 Tucson    15805.441373
Name: price, dtype: float64

# 16. 당뇨 여부 판단 데이터

In [87]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            3      102             74              0        0  29.5   
1            2      144             58             33      135  31.6   
2            5      136             82              0        0   0.0   
3           13      145             82             19      110  22.2   
4            1      117             60             23      106  33.8   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.121   32        0  
1                     0.422   25        1  
2                     0.640   69        0  
3                     0.245   57        0  
4                     0.466   27        0

**Question 102. Outcome 값에 따른 각 그룹의 각 컬럼의 평균 차이를 구하여라.**

In [88]:
# sol1
target = df.groupby('Outcome').mean()
abs(target.loc[0] - target.loc[1])

Pregnancies                  1.574159
Glucose                     29.943995
BloodPressure                3.202079
SkinThickness                2.952033
Insulin                     33.398645
BMI                          4.541437
DiabetesPedigreeFunction     0.143646
Age                          5.866939
dtype: float64

In [89]:
# sol2
df.groupby('Outcome').mean().diff().iloc[1, :] # loc[1] 

Pregnancies                  1.574159
Glucose                     29.943995
BloodPressure                3.202079
SkinThickness                2.952033
Insulin                     33.398645
BMI                          4.541437
DiabetesPedigreeFunction     0.143646
Age                          5.866939
Name: 1, dtype: float64

# 17. 넷플릭스 주식 데이터

In [90]:
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv")
df.head()

Date      Open      High       Low     Close  Adj Close     Volume
0  2002-05-23  1.156429  1.242857  1.145714  1.196429   1.196429  104790000
1  2002-05-24  1.214286  1.225000  1.197143  1.210000   1.210000   11104800
2  2002-05-28  1.213571  1.232143  1.157143  1.157143   1.157143    6609400
3  2002-05-29  1.164286  1.164286  1.085714  1.103571   1.103571    6757800
4  2002-05-30  1.107857  1.107857  1.071429  1.071429   1.071429   10154200

**Question 103. 매년 5월달의 open가격의 평균값을 데이터프레임으로 표현하라.**

In [91]:
# sol1
df['Date']= pd.to_datetime(df['Date'])
target = df[df['Date'].dt.month==5]
target.groupby(target['Date'].dt.strftime('%Y-%m'))['Open'].mean().to_frame('meanOpen')

meanOpen
Date               
2002-05    1.155833
2003-05    1.641497
2004-05    4.261143
2005-05    1.951905
2006-05    4.163571
2007-05    3.159351
2008-05    4.435034
2009-05    5.790571
2010-05   14.417071
2011-05   34.650272
2012-05   10.523247
2013-05   31.936429
2014-05   51.121292
2015-05   85.057429
2016-05   92.705715
2017-05  158.255455
2018-05  329.779541
2019-05  359.664548
2020-05  433.880499
2021-05  496.923996

In [92]:
# sol2
target = df.groupby(df['Date'].dt.strftime('%Y-%m')).mean(numeric_only=True)
answer = target.loc[target.index.str.contains('-05')].Open
pd.DataFrame(answer)

Open
Date               
2002-05    1.155833
2003-05    1.641497
2004-05    4.261143
2005-05    1.951905
2006-05    4.163571
2007-05    3.159351
2008-05    4.435034
2009-05    5.790571
2010-05   14.417071
2011-05   34.650272
2012-05   10.523247
2013-05   31.936429
2014-05   51.121292
2015-05   85.057429
2016-05   92.705715
2017-05  158.255455
2018-05  329.779541
2019-05  359.664548
2020-05  433.880499
2021-05  496.923996

# 18. NBA 선수 능력치 데이터

In [93]:
pd.set_option('display.max_columns',50)
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nba/nba.csv",encoding='latin',sep=';')
df.head()

Rk             Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%   3P  \
0   1   Precious Achiuwa   C   22  TOR  73  28  23.6  3.6   8.3  0.439  0.8   
1   2       Steven Adams   C   28  MEM  76  75  26.3  2.8   5.1  0.547  0.0   
2   3        Bam Adebayo   C   24  MIA  56  56  32.6  7.3  13.0  0.557  0.0   
3   4       Santi Aldama  PF   21  MEM  32   0  11.3  1.7   4.1  0.402  0.2   
4   5  LaMarcus Aldridge   C   36  BRK  47  12  22.3  5.4   9.7  0.550  0.3   

   3PA    3P%   2P   2PA    2P%   eFG%   FT  FTA    FT%  ORB  DRB   TRB  AST  \
0  2.1  0.359  2.9   6.1  0.468  0.486  1.1  1.8  0.595  2.0  4.5   6.5  1.1   
1  0.0  0.000  2.8   5.0  0.548  0.547  1.4  2.6  0.543  4.6  5.4  10.0  3.4   
2  0.1  0.000  7.3  12.9  0.562  0.557  4.6  6.1  0.753  2.4  7.6  10.1  3.4   
3  1.5  0.125  1.5   2.6  0.560  0.424  0.6  1.0  0.625  1.0  1.7   2.7  0.7   
4  1.0  0.304  5.1   8.8  0.578  0.566  1.9  2.2  0.873  1.6  3.9   5.5  0.9   

   STL  BLK  TOV   PF   PTS  
0  0.5  0.6  1.2  2.1   9.1  
1  0.9  0.8  1.5  2.0   6.9  
2  1.4  0.8  2.6  3.1  19.1  
3  0.2  0.3  0.5  1.1   4.1  
4  0.3  1.0  0.9  1.7  12.9

**Question 106. 선수들의 이름은 first_name+ 공백 + last_name으로 이루어져 있다. 가장 많은 first_name은 무엇이며 몇 회 발생하는지 확인하라.**

In [94]:
df['first_name'] = df['Player'].str.split(' ').str[0].str.lower()
df['first_name'].value_counts().head(1)

justin    14
Name: first_name, dtype: int64

**Question 109. G컬럼은 참여한 경기의 숫자이다. 팀별로 가장 높은 경기 참여 수를 가진 선수들의 경기 참여 숫자의 평균을 구하여라.**

In [95]:
# sol1
df.sort_values(['Tm', 'G'], ascending=False).drop_duplicates('Tm', keep='first')['G'].mean()

77.51612903225806

In [96]:
# sol2
df.groupby(['Tm', 'Player'])['G'].sum().reset_index().sort_values(by='G', ascending=False).drop_duplicates('Tm', keep='first').mean(numeric_only=True).values[0]

77.51612903225806

In [97]:
# sol3
df.sort_values(['Tm', 'G']).groupby('Tm').tail(1).G.mean()

77.51612903225806

# 참고

- [[TIL]DataFrame_ValueError: could not convert string to float: 'None' 해결](https://velog.io/@cbkyeong/TILDataFrameValueError-could-not-convert-string-to-float-None-%ED%95%B4%EA%B2%B0)